src/fparser/common/splitline.py

```diff
@@ -92,71 +94,120 @@ class ParenString(six.text_type):
 _is_name = re.compile(r'\w*\Z', re.I).match
 _is_simple_str = re.compile(r'\w*\Z', re.I).match
 _f2py_findall = re.compile(
-    r'(_F2PY_STRING_CONSTANT_\d+_|F2PY_EXPR_TUPLE_\d+)').findall
+    r'(_F2PY_STRING_CONSTANT_\d+_|F2PY_REAL_CONSTANT_\d+_|'
+    r'F2PY_EXPR_TUPLE_\d+)').findall
 
 
-class string_replace_dict(dict):
+class StringReplaceDict(dict):
     """
     Dictionary object that is callable for applying map returned
     by string_replace_map() function.
     """
     def __call__(self, line):
-        for k in _f2py_findall(line):
-            line = line.replace(k, self[k])
+        for key in _f2py_findall(line):
+            if key in self:
+                # We only replace the occurrence of `key` corresponding to
+                # the current result of the findall. This prevents the
+                # 'replace' also affecting subsequent matches that may
+                # have key as a substring (e.g. 'F2PY_EXPR_TUPLE_10'
+                # contains 'F2PY_EXPR_TUPLE_1').
+                line = line.replace(key, self[key], 1)
         return line
 
 
-def string_replace_map(line, lower=False,
-                       _cache={'index': 0, 'pindex': 0}):
+def string_replace_map(line, lower=False):
     """
     1) Replaces string constants with symbol `'_F2PY_STRING_CONSTANT_<index>_'`
     2) Replaces (expression) with symbol `(F2PY_EXPR_TUPLE_<index>)`
-    Returns a new line and the replacement map.
+    3) Replaces real numerical constants containing an exponent with symbol
+       `'F2PY_REAL_CONSTANT_<index>_'`
+
+    :param str line: the line of text in which to perform substitutions.
+    :param bool lower: whether or not the call to splitquote() should return \
+        items as lowercase (default is to leave the case unchanged).
+
+    :returns: a new line and the replacement map.
+    :rtype: 2-tuple of str and \
+            :py:class:`fparser.common.splitline.StringReplaceDict`
+
     """
+    # A valid exponential constant must begin with a digit or a '.' (and be
+    # preceeded by a non-'word' character or the start of the string).
+    exponential_constant = re.compile(
+        r"(?:[^\w]|^)(\d+[.]\d*|\d*[.]\d+|\d+)\s*[ed]\s*[+-]?\s*\d+(_\w*)?",
+        re.I)
+
+    str_idx = 0
+    const_idx = 0
+    parens_idx = 0
+
     items = []
-    string_map = string_replace_dict()
+    string_map = StringReplaceDict()
     rev_string_map = {}
     for item in splitquote(line, lower=lower)[0]:
         if isinstance(item, String) and not _is_simple_str(item[1:-1]):
             key = rev_string_map.get(item)
             if key is None:
-                _cache['index'] += 1
-                index = _cache['index']
-                key = "_F2PY_STRING_CONSTANT_%s_" % (index)
-                it = item[1:-1]
-                string_map[key] = it
-                rev_string_map[it] = key
+                str_idx += 1
```

Comment:
At first I just modified the way it was initialised because it was (intentionally I belatedly realised) keeping values for the duration of the parsing run. Then I realised we didn't need to keep the values. Then I realised that we didn't need to use a dictionary and thus we ended up here.

Reply:
That was a journey :-)